.copr/Makefile

```diff
@@ -1,23 +1,31 @@
 include Makefile
 
 outdir ?= $(PWD)/outdir
-spec ?= $(PWD)
+specdir ?= $(PWD)
 
 SOURCEDIR := $(shell rpmbuild -E %_sourcedir)
 SRCRPMDIR := $(shell rpmbuild -E %_srcrpmdir)
 
-srpm: SHORTNAME = $(shell basename $(PWD))
-srpm: LONGNAME = $(SHORTNAME)
-srpm: SERVICENAME = "yggdrasil"
-srpm: VERSION := $(shell echo $(VERSION) | awk -F. '{printf("%d.%d.%d", $$1, $$2, $$3+1)}')
-srpm: RELEASE = $(shell printf "0.%s.git.%s" $(shell git rev-list $(shell git describe --tags --abbrev=0 --always | tr -d '\n')..HEAD --count | tr -d '\n') $(shell git rev-parse --short HEAD | tr -d '\n'))
-srpm: PKGNAME = $(SHORTNAME)
+SHORTNAME = rhc
+LONGNAME = $(SHORTNAME)
+SERVICENAME = yggdrasil
+VERSION := $(shell echo $(VERSION) | awk -F. '{printf("%d.%d.%d", $$1, $$2, $$3+1)}')
+RELEASE = $(shell printf "0.%s.git.%s" $(shell git rev-list $(shell git describe --tags --abbrev=0 --always | tr -d '\n')..HEAD --count | tr -d '\n') $(shell git rev-parse --short HEAD | tr -d '\n'))
+PKGNAME = $(SHORTNAME)
 
-.PHONY: srpm
-srpm: deps
-	mkdir -p $(SOURCEDIR)
-	mkdir -p $(SRCRPMDIR)
-	mkdir -p $(outdir)
+# This should not be part of Makefile. COPR should provide some
+# way how to install packages required for running
+# make -f .copr/Makefile srpm outdir=./out
+.PHONY: deps
+deps:
+	rpm -qa > /tmp/rpm-list; \
+	grep '^golang-[0-9]' /tmp/rpm-list \
+	&& grep '^git-[0-9]' /tmp/rpm-list \
+	&& grep '^rpm-build-[0-9]' /tmp/rpm-list \
+	|| dnf install -y golang git rpm-build
+
```

Comment:
While I understand the idea behind this, I'd rather keep `make srpm` for copr only, and thus here unconditionally installing the needed packages with `dnf`. Overloading `make srpm` with other meanings than what copr wants means complicating this `Makefile`, and the code above shows exactly this.

With the new split targets for the spec file and the tarball, now there is less need to use `make srpm` manually. If it is wanted to have also an easy way to generate the srpm, I'd simply add a new target:
```make
source-rpm: spec tarball
    rpmbuild -bs $(specdir)/rhc.spec

srpm: deps source-srpm
    ...
```
this would have the advantages of reusing the existing step (part of the `make srpm` flow for copr), and not overload it with other meanings. Bonus point: using this new target would not bloat your local "SOURCES" and "SRPMS" directories of rpmbuild.

Reply:
When there is empty `%changelog` or no `%changelog` in `rhc.spec` file, then warning is printed during linting of rpm.

I don't like idea of having two targets for building `src.rpm`. Current solution just works. `make srpm` makes `src.rpm`. There is no overloading of meaning.

Reply:
> When there is empty `%changelog` or no `%changelog` in `rhc.spec` file, then warning is printed during linting of rpm.

That warning is harmless. If that is a concern, I'd rather use `%autochangelog` instead; please do not store a static changelog that is useless and gets obsolete soon.

> I don't like idea of having two targets for building `src.rpm`. Current solution just works. `make srpm` makes `src.rpm`.

`make srpm` also copies the bits to the "SOURCES" and "SRPMS" directories of rpmbuild (see the output of `rpmbuild -E %_sourcedir` and `rpmbuild -E %_srcrpmdir`), because that is the "interface" of `make srpm` for copr. Copying them for a local run is wasteful, because it fills up your own local directories with tarballs and srpm files.

> There is no overloading of meaning.

There is: now `make srpm` has to work also for what was not designed (i.e. build in copr).

Reply:
> I really don't care about copying files to "SOURCES" and "SRPMS".

That isn't really a reason to silently fill up some place that you might not even know about.

> We all can use `rm` command.

Let's go by your logic then: we all can use `rpmbuild` command.

> Does `make srpm` work? It does. There is no problem.

`make srpm` is for copr, please leave it to copr exclusively without overloading its function. The fact that it "works for you" does not mean it is a good idea in general or there is no problem with it. Implementing an hack and saying that "it works" is not a good way forward.

Please drop this hack, thanks, as already asked.

Reply:
The `make srpm` is common target in other projects e.g. https://github.com/candlepin/virt-who/blob/main/Makefile#L16. The necessity to run `make srpm` as a root is a hack. The necessity of installing any RPM as a side effect of `srpm` target is a hack. I try to make it only less painful.

BTW: I would say that you are not contributor to this project and your are not in the position to request or enforce anything. Your advice is welcome, but the style of communication you are showing in this PR is not acceptable.

Reply:
The `srpm` target in the file `.copr/Makefile` is an interface for building an SRPM required by COPR. Hypothetically, if COPR were to change the required output of the `make srpm` interface, we would be compelled to update the `srpm` target follow the new expectations, or our system would stop working with COPR. To me, that means that `make srpm` is not a general-purpose "build a source RPM" make recipe. Instead, it is a COPR-specific interface. Additionally, I think COPR assumes that it always invokes `make -f .copr/Makefile srpm` as root. One more reason why this interface is very COPR-specific.

We would be wise to abstract ourselves away from having to comply with changes that might arise due to that external API. (I did this in yggdrasil; a general-purpose SRPM build recipe exists in the `dist/srpm` directory, and COPR's `.copr/Makefile` interface uses it.) If we want to support a general-purpose SRPM build recipe, it should be in `./Makefile` instead of `.copr/Makefile`. virt-who appears to do it this way; it has a general, top-level `Makefile` with a 'srpm' target that invokes `tito` to compile a source RPM.

Reply:
COPR can assumes that `make -f .copr/Makefile srpm` is invoked by root. Current change will not influence this assumption.

If COPR build was broken, then fixing/testing would be easier with this change... It would allow to trigger `make -f .copr/Makefile srpm` on local machine as non-root user. I think that it is better to try to fix/test broken COPR build locally as non-root user, then running something like this as root, or re-trigger COPR build again and again after change is pushed to some PR.

I aware that this change is not perfect... I would like to improve it (e.g. it would be good to have the list of required packages in some variable and do not duplicate code), but first I need to know that something like this is wanted.

Reply:
Good points. I agree that it does make debugging the COPR makefile and target easier than it is now.